Transactions.md

```diff
@@ -1,182 +1,292 @@
-  * Fee Rate: `FeeUnit`
-
-Where
-  - `Local Address` is any unused local address
-  - `Remote's ToLocalOutput` is of the form [specified above](#CETOutputs)
-### <a name="ClosingInputs">Inputs</a>
-  * Input Spending(P2WSH(Remote's ToLocalOutput))
-### <a name="ClosingOutputs">Outputs</a>
-  * One output corresponding to `LocalAddress` with value `P2WSH(Remote's ToLocalOutput).value - fee`
+This details the exact format of all on-chain transactions, which both sides need to agree on to ensure signatures are valid.
 
+# Table of Contents
+
+* [Funding Transaction](#funding-transaction)
+  * [Funding Transaction Input and Output Ordering](#funding-transaction-input-and-output-ordering)
+  * [Funding Inputs](#funding-inputs)
+  * [FundingOutput](#funding-output)
+  * [Change Outputs](#change-outputs)
+* [Contract Execution Transaction](#contract-execution-transaction)
+  * [CET Outputs](#cet-outputs)
+    * [Offerer Output](#offerer-output)
+    * [Accepter Output](#accepter-output)
+* [Refund Transaction](#refund-transaction)
+* [Fees](#fees)
+  * [Expected Weight of the Funding Transaction](#expected-weight-of-the-funding-transaction)
+  * [Expected Weight of the Contract Execution or Refund Transaction](#expected-weight-of-the-contract-execution-or-refund-transaction)
+* [Test Vectors](#test-vectors)
+* [References](#references)
+* [Authors](#authors)
+
+# Funding Transaction
+
+* version: 2
+* locktime: 0
+
+The funding inputs and change output script public keys are negotiated in the offer and accept messages.
+
+## Funding Transaction Input and Output Ordering
+
+The Offerer/Initiator's inputs come first and in the order that they were listed in the Offer message, the Accepter's inputs come after this in the order they are listed in the Accept message. The funding output comes first, followed by the Offerer's change output if applicable, and then the Accepter's change output if applicable.
+
+Note that this will likely change in the future.
+
+## Funding Inputs
+
+All funding inputs must be Segwit or nested P2SH(Segwit) in order to protect against malleability attacks.
+
+## Funding Output
+
+* The funding output script is a P2WSH to:
+
+```
+2 <offer_funding_pubkey> <accept_funding_pubkey> 2 OP_CHECKMULTISIG
+```
+
+* Where both `pub_key`s are in compressed format.
+
+## Change Outputs
+
+The funding transaction's change outputs should pay to the address specified in the relevant offer/accept message. A change output's value should equal the total funding amount of that party subtracted by their total collateral as well as their fees for both this transaction as well as their fees for the largest possible Contract Execution Transaction. If this value is below the dust limit of `1000 satoshis`, then that party must include additional funding in order to ensure they have a valid anchor output.
+
+# Contract Execution Transaction
+
+Also known as a CET.
+
+* version: 2
+* locktime: `contract_maturity`
```

Comment:
So perhaps this field needs to be renamed but I strongly believe it must be included.

I agree that the decision of what this value should be should happen at the application layer but it definitely still needs to be specified here. This value can be set to 0 for example. It can be set to the current block height, it can be set to the time at which you are expected to be back online. Regardless, it definitely must be specified to avoid ambiguity.

It doesn't provide any guarantees about the oracle but it certainly gives parties (should they set this) time to cooperate to change oracles if an oracle becomes compromised.

I don't believe that this time-lock is associated with fee concerns.

Reply:
>This value can be set to 0 for example. It can be set to the current block height, it can be set to the time at which you are expected to be back online. Regardless, it definitely must be specified to avoid ambiguity.

Ok, I agree with you on the need for it to be specified, I think it just offers faux security guarantees if we start setting it to be value in the future. It makes it harder for us as protocol developers to test, and doesn't add any security guarantees to the end user because once the oracle publishes signatures related to the event your leverage is gone.

Reply:
I'm with @Christewart on this one as I tried to express in https://github.com/discreetlogcontracts/dlcspecs/pull/59#discussion_r470360648 It's a real pain for testing, can affect user experience (in many cases you might not be able to broadcast your TX even after the oracle signature is vailable_ and I agree it doesn't provide anything valuable:

> It doesn't provide any guarantees about the oracle but it certainly gives parties (should they set this) time to cooperate to change oracles if an oracle becomes compromised.

I don't see that as a valid argument, if both parties are cooperative and/or honest, they can change oracle. If one of the party thinks the compromise of the oracle is favorable to them, they'll just take advantage of it whether there is a time lock or not, the time lock really doesn't change anything.

If the point is to set it in the past or to current block height I'm ok with it, but then I feel we should specify how that should work in the protocol (e.g. which values a client should accept or reject).

Reply:
A non-mature transaction isn't going to propagate (where maturity is defined as relative/absolute timelock not expiring at next block). If your transaction isn't in network mempools you can't CPFP it. So it may complexify a bit your broadcast logic as you might learn a valid witness (oracle scalar) to authorize a spend but have to cache broadcast until reaching unix/block height.

I don't think it provides any security for the reason mentioned by Tibo, unless you can get with a proof of oracle compromise and thus active immediately the refund path. Security with regards to oracle compromise is interesting to think of but beyond current convervsation IMO.

Wheres it does provide setting this value to non-far-from-current-blockheight is both as a anti-fee snipping mechanism and avoiding a fingerprint. Bitcoin Core wallet implements fingeprint but set back from few blocks ~10% of transactions to avoid fingerprinting pre-signed Coinjoins to account for slow multi-party signing rounds. It would be great if we can mask DLC in this set. We don't care about the anti-fee snipping as our fee model is likely to be a just-in-time CPFP and won't be interesting for a miner to fee snip a poor-feerate parent.